File: MarketPortal/Market/Data/Models/Offer.cs
using Microsoft.AspNetCore.Mvc;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Market.Data.Models
{
    public class Offer
    {
        [Key]
        public int Id { get; set; }

        [Required]
        [StringLength(28)]
        public string Title { get; set; }

        [Required]
        [StringLength(16)]
        public string Town { get; set; }

        [Required]
        [StringLength(300)]
        public string Description { get; set; }

        public decimal AvgRating { get; set; }
        

        [Required]
        public decimal PricePerKG { get; set; }
        [Required]
        [ForeignKey(nameof(User))]
        public Guid OwnerId { get; set; }
        //public User Owner { get; set; }

        [Required]
        [ForeignKey(nameof(Stock))]
        public int StockId { get; set; }

        public virtual List<Review> Reviews { get; set; } = new List<Review>();


        public Stock Stock { get; set; }

        [Required]
        public DateTime DatePosted { get; set; }

        [Required]
        [Range(minimum: 5, maximum: 20)]
        public int Discount { get; set; }

        [ForeignKey(nameof(AdvertiseSettings))]
        public int? AdvertiseSettingsId { get; set; }

        public virtual AdvertiseSettings? AdvertiseSettings { get; set; }

    }
}

File: MarketFlutter/lib/models/offer.dart


import 'dart:convert';

import 'package:market/models/review.dart';
import 'package:market/models/stock.dart';

class Offer {
  // public int Id { get; set; }
  // public string Title { get; set; }
  // public double PricePerKG { get; set; }
  // public Guid OwnerId { get; set; }
  // public User Owner { get; set; }
  // public bool inSeason { get; set; }
  // public int OfferTypeId { get; set; }
  // public OfferType OfferType { get; set; }


  int id;
  String title;
  String town;
  String description;
  double pricePerKG;
  String ownerId;
  int stockId;
  int offerTypeId;
  String offerTypeName;
  DateTime datePosted;
  Stock stock;
  List<Review>? reviews;
  double avgRating;
  // Constructor
  Offer({required this.id, required this.title, required this.town, required this.pricePerKG, required this.ownerId, required this.stockId, required this.description, required this.avgRating, required this.datePosted, required this.offerTypeId, required this.stock, this.reviews, required this.offerTypeName});

  factory Offer.fromJson(Map<String, dynamic> json) {
    print(json);
    Offer res = Offer(
      id: json['id'] as int,
      title: json['title'] as String,
      town: json['town'] as String,
      description: json['description'] as String,
      pricePerKG: json['pricePerKG']+.0,
      ownerId: json['ownerId'] as String,
      datePosted: DateTime.parse(json['datePosted']),
      stockId: json['stockId'] as int, //TODO: fix stock DTO
      offerTypeId: json['stock']['offerTypeId'] as int,
      avgRating: json['avgRating']+.0,
      offerTypeName: json['offerType'] ?? json['stock']['offerType']['name'] as String,
      stock: Stock.fromJson(json['stock']),
      //reviews: List<Review>.from(json['reviews'].map((model)=> Review.fromJson(model))),
    );
    return res;
  }

  // Method to convert User instance to a JSON map
  Map<String, dynamic> toJson() {
    //TODO: Add all types when fix api
    return {
      'id': id,
      'title': title,
      'pricePerKG': pricePerKG,
      'ownerId': ownerId,
      'stockId': stockId,
      'description': description,
      'town': town,
      'datePosted': datePosted.toString(),
      'reviews': reviews?.map((x) => x.toJson()).toList(),
      'stock': stock.toJson(),
      'avgRating': avgRating,
    };
  }
}